usr.bin/ctags/ctags.1
.\" Copyright (c) 1987, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ctags.1	8.1 (Berkeley) 6/6/93
.\"
.TH CTAGS 1 "October 1997" GNO "Commands and Applications"
.SH NAME
.BR ctags
\- create a tags file
.SH SYNOPSIS
.BR ctags
.RB [ -BFadtuwvx ]
.RB [ -f 
.IR tagsfile ]
.IR name ...
.SH DESCRIPTION
.BR Ctags
makes a tags file for
.BR ex (1)
from the specified C, Pascal, Fortran, YACC, lex, and lisp sources.
A tags file gives the locations of specified objects in a group of files.
Each line of the tags file contains the object name, the file in which it
is defined, and a search pattern for the object definition, separated by
white-space.
Using the
.BR tags
file,
.BR ex (1)
can quickly locate these object definitions.
Depending upon the options provided to
.BR ctags ,
objects will consist of subroutines, typedefs, defines, structs,
enums and unions.
.IP "\fB-B\fR"
use backward searching patterns
.RB ( "?...?" ).
.IP "\fB-F\fR"
use forward searching patterns
.RB ( "/.../" )
(the default).
.IP "\fB-a\fR"
append to
.BR tags
file.
.IP "\fB-d\fR"
create tags for #defines that don't take arguments; #defines
that take arguments are tagged automatically.
.IP "\fB-f\fR"
Places the tag descriptions in a file called
.IR tagsfile .
The default behaviour is to place them in a file called
.BR tags .
.IP "\fB-t\fR"
create tags for typedefs, structs, unions, and enums.
.IP "\fB-u\fR"
update the specified files in the
.BR tags
file, that is, all
references to them are deleted, and the new values are appended to the
file.  (Beware: this option is implemented in a way which is rather
slow; it is usually faster to simply rebuild the
.BR tags
file.)
.IP "\fB-v\fR"
An index of the form expected by
.BR vgrind (1)
is produced on the standard output.  This listing
contains the object name, file name, and page number (assuming 64
line pages).  Since the output will be sorted into lexicographic order,
it may be desired to run the output through
.BR sort (1).
Sample use:
.nf

		ctags \-v files \&| sort \-f > index
		vgrind \-x index
.fi
.IP "\fB-w\fR"
suppress warning diagnostics.
.IP "\fB-x\fR"
.BR ctags
produces a list of object
names, the line number and file name on which each is defined, as well
as the text of that line and prints this on the standard output.  This
is a simple index which can be printed out as an off-line readable
function index.
.LP
Files whose names end in
.BR \&.c
or
.BR \&.h
are assumed to be C
source files and are searched for C style routine and macro definitions.
Files whose names end in
.BR \&.y
are assumed to be YACC source files.
Files whose names end in
.BR \&.l
are assumed to be lisp files if their
first non-blank character is `;', `(', or `[',
otherwise, they are
treated as lex files.  Other files are first examined to see if they
contain any Pascal or Fortran routine definitions, and, if not, are
searched for C style definitions.
.LP
The tag
.BR main
is treated specially in C programs.  The tag formed
is created by prepending
.BR M
to the name of the file, with the
trailing
.BR \&.c
and any leading pathname components removed.  This
makes use of
.BR ctags
practical in directories with more than one
program.
.LP
Yacc and lex files each have a special tag.
.BR Yyparse
is the start
of the second section of the yacc file, and
.BR yylex
is the start of
the second section of the lex file.
.SH FILES
.IP "\fBtags\fR"
default output tags file
.SH DIAGNOSTICS
.BR Ctags
exits with a value of 1 if an error occurred, 0 otherwise.
Duplicate objects are not considered errors.
.SH ATTRIBUTIONS
This command was ported from FreeBSD source code
for distribution with GNO/ME 2.0.6.
.SH SEE ALSO
.BR ex (1),
.BR vi (1)
.SH BUGS
Recognition of
.BR functions  ,
.BR subroutines
and
.BR procedures
for FORTRAN
and Pascal is done is a very simpleminded way.  No attempt
is made to deal with block structure; if you have two Pascal procedures
in different blocks with the same name you lose.
.BR Ctags
doesn't
understand about Pascal types.
.LP
The method of deciding whether to look for C, Pascal or FORTRAN
functions is a hack.
.LP
.BR Ctags
relies on the input being well formed, and any syntactical
errors will completely confuse it.  It also finds some legal syntax
confusing; for example, since it doesn't understand #ifdef's
(incidentally, that's a feature, not a bug), any code with unbalanced
braces inside #ifdef's
will cause it to become somewhat disoriented.
In a similar fashion, multiple line changes within a definition will
cause it to enter the last line of the object, rather than the first, as
the searching pattern.  The last line of multiple line typedef's
will similarly be noted.
.SH HISTORY
The
.BR Ctags
command appeared in 3.0BSD.


usr.bin/ctags/ctags.c
/*
 * Copyright (c) 1987, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1987, 1993, 1994\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)ctags.c	8.3 (Berkeley) 4/2/94";
#endif /* not lint */

#include <err.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

#include "ctags.h"

/*
 * ctags: create a tags file
 */

NODE	*head;			/* head of the sorted binary tree */

				/* boolean "func" (see init()) */
bool	_wht[256], _etk[256], _itk[256], _btk[256], _gd[256];

FILE	*inf;			/* ioptr for current input file */
FILE	*outf;			/* ioptr for tags file */

long	lineftell;		/* ftell after getc( inf ) == '\n' */

int	lineno;			/* line number of current line */
int	dflag;			/* -d: non-macro defines */
int	tflag;			/* -t: create tags for typedefs */
int	vflag;			/* -v: vgrind style index output */
int	wflag;			/* -w: suppress warnings */
int	xflag;			/* -x: cxref style output */

char	*curfile;		/* current input file name */
char	searchar = '/';		/* use /.../ searches by default */
char	lbuf[LINE_MAX];

void	init __P((void));
void	find_entries __P((char *));

int
main(int argc, char **argv)
{
static char	*outfile = "tags";	/* output file */
int	aflag;				/* -a: append to tags */
int	uflag;				/* -u: update tags */
int	exit_val;			/* exit value */
int	step;				/* step through args */
int	ch;				/* getopts char */
static char	cmd[100];			/* too ugly to explain */

#ifdef __ORCAC203__
	aflag = NO; uflag = NO;
#else
	aflag = uflag = NO;
#endif
	while ((ch = getopt(argc, argv, "BFadf:tuwvx")) != EOF)
		switch(ch) {
		case 'B':
			searchar = '?';
			break;
		case 'F':
			searchar = '/';
			break;
		case 'a':
			aflag++;
			break;
		case 'd':
			dflag++;
			break;
		case 'f':
			outfile = optarg;
			break;
		case 't':
			tflag++;
			break;
		case 'u':
			uflag++;
			break;
		case 'w':
			wflag++;
			break;
		case 'v':
			vflag++;
		case 'x':
			xflag++;
			break;
		case '?':
		default:
			goto usage;
		}
	argv += optind;
	argc -= optind;
	if (!argc) {
usage:		(void)fprintf(stderr,
			"usage: ctags [-BFadtuwvx] [-f tagsfile] file ...\n");
		exit(1);
	}

	init();

#ifdef __ORCAC203__
	for (exit_val = 0, step = 0; step < argc; ++step)
#else
	for (exit_val = step = 0; step < argc; ++step)
#endif
		if (!(inf = fopen(argv[step], "r"))) {
			warn("%s", argv[step]);
			exit_val = 1;
		}
		else {
			curfile = argv[step];
			find_entries(argv[step]);
			(void)fclose(inf);
		}

	if (head)
		if (xflag)
			put_entries(head);
		else {
			if (uflag) {
				for (step = 0; step < argc; step++) {
					(void)sprintf(cmd,
						"mv %s OTAGS; fgrep -v '\t%s\t' OTAGS >%s; rm OTAGS",
							outfile, argv[step],
							outfile);
					system(cmd);
				}
				++aflag;
			}
			if (!(outf = fopen(outfile, aflag ? "a" : "w")))
				err(exit_val, "%s", outfile);
			put_entries(head);
			(void)fclose(outf);
			if (uflag) {
				(void)sprintf(cmd, "sort -o %s %s",
						outfile, outfile);
				system(cmd);
			}
		}
	exit(exit_val);
}

/*
 * init --
 *	this routine sets up the boolean psuedo-functions which work by
 *	setting boolean flags dependent upon the corresponding character.
 *	Every char which is NOT in that string is false with respect to
 *	the pseudo-function.  Therefore, all of the array "_wht" is NO
 *	by default and then the elements subscripted by the chars in
 *	CWHITE are set to YES.  Thus, "_wht" of a char is YES if it is in
 *	the string CWHITE, else NO.
 */
void
init(void)
{
int		i;
unsigned char	*sp;

	for (i = 0; i < 256; i++) {
#ifdef __ORCAC203__
		_wht[i] = NO; _etk[i] = NO; _itk[i] = NO; _btk[i] = NO;
#else
		_wht[i] = _etk[i] = _itk[i] = _btk[i] = NO;
#endif
		_gd[i] = YES;
	}
#define	CWHITE	" \f\t\n"
	for (sp = CWHITE; *sp; sp++)	/* white space chars */
		_wht[*sp] = YES;
#define	CTOKEN	" \t\n\"'#()[]{}=-+%*/&|^~!<>;,.:?"
	for (sp = CTOKEN; *sp; sp++)	/* token ending chars */
		_etk[*sp] = YES;
#define	CINTOK	"ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz0123456789"
	for (sp = CINTOK; *sp; sp++)	/* valid in-token chars */
		_itk[*sp] = YES;
#define	CBEGIN	"ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"
	for (sp = CBEGIN; *sp; sp++)	/* token starting chars */
		_btk[*sp] = YES;
#define	CNOTGD	",;"
	for (sp = CNOTGD; *sp; sp++)	/* invalid after-function chars */
		_gd[*sp] = NO;
}

/*
 * find_entries --
 *	this routine opens the specified file and calls the function
 *	which searches the file.
 */
void
find_entries(char *file)
{
char	*cp;

	lineno = 0;				/* should be 1 ?? KB */
	if (cp = strrchr(file, '.')) {
		if (cp[1] == 'l' && !cp[2]) {
			int	c;

			for (;;) {
				if (GETC(==, EOF))
					return;
				if (!iswhite(c)) {
					rewind(inf);
					break;
				}
			}
#define	LISPCHR	";(["
/* lisp */		if (strchr(LISPCHR, c)) {
				l_entries();
				return;
			}
/* lex */		else {
				/*
				 * we search all 3 parts of a lex file
				 * for C references.  This may be wrong.
				 */
				toss_yysec();
				(void)strcpy(lbuf, "%%$");
				pfnote("yylex", lineno);
				rewind(inf);
			}
		}
/* yacc */	else if (cp[1] == 'y' && !cp[2]) {
			/*
			 * we search only the 3rd part of a yacc file
			 * for C references.  This may be wrong.
			 */
			toss_yysec();
			(void)strcpy(lbuf, "%%$");
			pfnote("yyparse", lineno);
			y_entries();
		}
/* fortran */	else if ((cp[1] != 'c' && cp[1] != 'h') && !cp[2]) {
			if (PF_funcs())
				return;
			rewind(inf);
		}
	}
/* C */	c_entries();
}

